libtenzir/builtins/operators/where_map.cpp

```diff
@@ -223,30 +229,181 @@ class where_operator2 final : public crtp_operator<where_operator2> {
   bool warn_;
 };
 
-class plugin2 final : public virtual operator_plugin2<where_operator2> {
-public:
-  explicit plugin2(bool warn) : warn_{warn} {
+TENZIR_ENUM(mode, map, where);
+
+struct arguments {
+  ast::expression field;
+  ast::simple_selector capture;
+  ast::expression expr;
+};
+
+auto make_where_map_function(function_plugin::invocation inv, session ctx,
+                             enum mode mode) -> failure_or<function_ptr> {
+  auto args = arguments{};
+  TRY(argument_parser2::function(fmt::to_string(mode))
+        .add(args.field, "<field>")
+        .add(args.capture, "<capture>")
+        .add(args.expr, "<expr>")
+        .parse(inv, ctx));
+  // We internally use the empty string for our top-level dummy field, so it
+  // must not be used in the capture name.
+  if (args.capture.has_this() and not args.capture.path().empty()
+      and args.capture.path().front().name.empty()) {
+    diagnostic::error("capture name must not start with an empty string")
+      .primary(args.capture.path().front().location)
+      .emit(ctx);
+    return failure::promise();
   }
+  return function_use::make(
+    [mode, args = std::move(args)](function_plugin::evaluator eval,
+                                   session ctx) -> series {
+      auto field = eval(args.field);
+      if (field.as<null_type>()) {
+        return field;
+      }
+      auto field_list = field.as<list_type>();
+      if (not field_list) {
+        diagnostic::error("expected `list`, but got `{}`", field.type.kind())
+          .primary(args.field)
+          .emit(ctx);
+        return series::null(null_type{}, eval.length());
+      }
+      // We get the schema name from the parent evsaluator so that we can make
+      // @name available in the mapped expression.
+      const auto name
+        = eval(ast::meta{ast::meta::name, location::unknown}).as<string_type>();
+      TENZIR_ASSERT(name);
+      TENZIR_ASSERT(name->length() > 0);
+      TENZIR_ASSERT(name->array->IsValid(0));
+      // We get the field's inner values array and create a dummy table slice
+      // with a single field to evaluate the mapped expression on.
+      auto values
+        = series{field_list->type.value_type(), field_list->array->values()};
+      auto slice = table_slice{
+        check(arrow::RecordBatch::FromStructArray(
+          std::make_shared<arrow::StructArray>(
+            arrow::struct_({{"", arrow::null()}}), values.length(),
+            std::vector{
+              check(arrow::MakeArrayOfNull(arrow::null(), values.length()))}))),
+        type{name->array->GetView(0), record_type{{"", null_type{}}}},
+      };
+      slice = assign(args.capture, values, slice, ctx);
+      TENZIR_ASSERT(as<record_type>(slice.schema()).num_fields() == 2);
+      slice = transform_columns(
+        slice, {{offset{0},
+                 [](struct record_type::field, std::shared_ptr<arrow::Array>) {
+                   return indexed_transformation::result_type{};
+                 }}});
```

Comment:
This whole dance is only necessary because `assign` cannot work with empty table slices. @jachris do you know if that'd be hard to change?

Reply:
You mean a table slice with 0 rows? I don't see anything in `assign` that would fail for that. How does "cannot work" manifest itself?

Reply:
I mean one with 0 columns and through that an arbitrary length. Converting a table slice to a record batch fails if there are no columns.

Reply:
`from {} | x = 42`, which ends up calling `to_record_batch` in `assign`, works for me. Sorry, I'm not sure what you mean. Are you perhaps passing a default-initialized table slice?

Reply:
Ah, that was my mistake. Fixed it now without needing the dummy field.